peps/pep-0747.rst

```diff
@@ -0,0 +1,903 @@
+expression objects. Therefore ``TypeExpr`` is proposed as new notation
+instead.
+
+For a longer explanation of why we don’t just widen ``type[T]`` to
+accept all type expressions, see
+:ref:`widen_type_C_to_support_all_type_expressions`.
+
+
+Common kinds of functions that would benefit from TypeExpr
+----------------------------------------------------------
+
+`A survey of various Python libraries`_ revealed a few kinds of commonly
+defined functions which would benefit from ``TypeExpr[]``:
+
+.. _A survey of various Python libraries: https://github.com/python/mypy/issues/9773#issuecomment-2017998886
+
+-  Assignability checkers:
+
+   -  Returns whether a value is assignable to a type expression. If so
+      then also narrows the type of the value to match the type
+      expression.
+   -  Pattern 1:
+      ``def isassignable[T](value: object, typx: TypeExpr[T]) -> TypeIs[T]``
+   -  Pattern 2:
+      ``def ismatch[T](value: object, typx: TypeExpr[T]) -> TypeGuard[T]``
+   -  Examples: beartype.\ `is_bearable`_, trycast.\ `isassignable`_,
+      typeguard.\ `check_type`_, xdsl.\ `isa`_
+
+.. _is_bearable: https://github.com/beartype/beartype/issues/255
+.. _isassignable: https://github.com/davidfstr/trycast?tab=readme-ov-file#isassignable-api
+.. _check_type: https://typeguard.readthedocs.io/en/latest/api.html#typeguard.check_type
+.. _isa: https://github.com/xdslproject/xdsl/blob/ac12c9ab0d64618475efb98d1d197bdd79f593c3/xdsl/utils/hints.py#L23
+
+-  Converters:
+
+   -  If a value is assignable to (or coercible to) a type expression
+      then returns the value narrowed to (or coerced to) that type
+      expression. Otherwise raises an exception.
+   -  Pattern 1:
+      ``def convert[T](value: object, typx: TypeExpr[T]) -> T``
+
+      -  Examples: cattrs.BaseConverter.\ `structure`_, trycast.\ `checkcast`_,
+         typedload.\ `load`_
+
+   -  Pattern 2:
+
+      ::
+
+        class Converter[T]:
+            def __init__(self, typx: TypeExpr[T]) -> None: ...
+            def convert(self, value: object) -> T: ...
+
+      -  Examples: pydantic.\ `TypeAdapter(T).validate_python`_,
+         mashumaro.\ `JSONDecoder(T).decode`_
+
+.. _structure: https://github.com/python-attrs/cattrs/blob/5f5c11627a7f67a23d6212bc7df9f96243c62dc5/src/cattrs/converters.py#L332-L334
+.. _checkcast: https://github.com/davidfstr/trycast#checkcast-api
+.. _load: https://ltworf.github.io/typedload/
+.. _TypeAdapter(T).validate_python: https://stackoverflow.com/a/61021183/604063
+.. _JSONDecoder(T).decode: https://github.com/Fatal1ty/mashumaro?tab=readme-ov-file#usage-example
+
+-  Typed field definitions:
+
+   -  Pattern:
+
+      ::
+
+        class Field[T]:
+            value_type: TypeExpr[T]
+
+   -  Examples: attrs.\ `make_class`_, dataclasses.\ `make_dataclass`_, `openapify`_
+
+.. _make_class: https://www.attrs.org/en/stable/api.html#attrs.make_class
+.. _make_dataclass: https://github.com/python/typeshed/issues/11653
```

Comment:
Wouldn't this need an annotation expression instead of a type expression?

Reply:
I don't think so. None of the "annotation expression but not type expression" forms - with 1 exception† - make sense to use as field types: (Not)Required, ReadOnly, ClassVar, Final, InitVar† (🤔), Unpack.

† InitVar is the only problematic case.

I'll probably need to think about InitVar some more...

Reply:
I added a footnote talking about `InitVar` being a special case with `dataclasses.make_dataclass`